### PR TITLE
feat: allow including/excluding collection when adding datasources

### PR DESCRIPTION
**Pull request by romain-gilliotte**
### General

- [ ] Write an explicit title for the Pull Request, following [Conventional Commits specification](https://www.conventionalcommits.org)
- [ ] Test manually the implemented changes
- [ ] Validate the code quality (indentation, syntax, style, simplicity, readability)

### Security

- [ ] Consider the security impact of the changes made

### CODE REVIEW

File: packages/datasource-customizer/src/decorators/publication-collection/datasource.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import { DataSource } from '@forestadmin/datasource-toolkit';
+import DataSourceDecorator from '../datasource-decorator';
+import PublicationFieldCollectionDecorator from '../publication-field/collection';
+
+export default class PublicationCollectionDataSourceDecorator extends DataSourceDecorator<PublicationFieldCollectionDecorator> {
```
Comment (SteveBunlon) [inline packages/datasource-customizer/src/decorators/publication-collection/datasource.ts:5]: I don't understand the choice of `PublicationCollectionDataSourceDecorator` can you explain please ? (I mean, the naming)

File: packages/datasource-customizer/src/decorators/publication-collection/datasource.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    // Delete the relations to the collections we're not keeping.
+    for (const collection of Object.values(this._collections)) {
+      for (const [fieldName, field] of Object.entries(collection.schema.fields)) {
+        if (
+          (field.type !== 'Column' && deleted.has(field.foreignCollection)) ||
+          (field.type === 'ManyToMany' && deleted.has(field.throughCollection))
+        )
+          collection.changeFieldVisibility(fieldName, false);
```
Comment (SteveBunlon) [inline packages/datasource-customizer/src/decorators/publication-collection/datasource.ts:29]: Is there a reason why we simply set the visibility to false, and not delete it entirely ?

File: packages/datasource-customizer/src/datasource-customizer.ts [hunk 3]
```diff
@@ -36,11 +37,21 @@ export default class DataSourceCustomizer<S extends TSchema = TSchema> {
...
   addDataSource(factory: DataSourceFactory, options?: DataSourceOptions): this {
     this.customizations.push(async logger => {
-      const dataSource = await factory(logger);
-      const renamedDecorator = new RenameCollectionDataSourceDecorator(dataSource);
-      renamedDecorator.renameCollections(options?.rename);
+      let dataSource = await factory(logger);
 
-      this.compositeDataSource.addDataSource(renamedDecorator);
+      if (options?.include || options?.exclude) {
+        const publicationDecorator = new PublicateCollectionDataSourceDecorator(dataSource);
```
Comment (SteveBunlon) [inline packages/datasource-customizer/src/datasource-customizer.ts:43]: I don't understand Publicate here

File: docs/datasources/connection/relationships.md [hunk 4]
```diff
@@ -1,29 +1,26 @@
+Adding as many datasources as you need is good, inter-connecting them together is better.
```
Comment (SteveBunlon) [inline docs/datasources/connection/relationships.md:1]: This is maybe a change you had in mind before this PR. Just to point this out, this does not look to be related to this PR. Is this change intended ?

File: packages/datasource-customizer/src/decorators/publication-collection/datasource.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    this.validateCollectionName([...(include ?? []), ...(exclude ?? [])]);
+
+    const deleted = new Set<string>();
+
+    // List collection we're keeping from the white/black list.
+    for (const name of Object.keys(this._collections)) {
+      if ((include && !include.includes(name)) || exclude?.includes(name)) {
+        deleted.add(name);
+      }
+    }
```
Comment (SteveBunlon) [inline packages/datasource-customizer/src/decorators/publication-collection/datasource.ts:20]: This means that exclude and include can both be used ?

File: packages/datasource-customizer/test/decorators/publication-collection/datasource.test.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        'Unknown collection name: "unknown"',
+      );
+    });
+
+    it('should be able to remove "librariesBooks" collection', () => {
+      decoratedDataSource.publicate(['libraries', 'books']);
+
+      expect(() => decoratedDataSource.getCollection('librariesBooks')).toThrow();
+      expect(decoratedDataSource.getCollection('libraries').schema.fields).not.toHaveProperty(
+        'myLibrary',
```
Comment (SteveBunlon) [inline packages/datasource-customizer/test/decorators/publication-collection/datasource.test.ts:95]: Shouldn't it be `myBooks` here ?

**Review by SteveBunlon — CHANGES_REQUESTED** [review]
Could be great to have some tests with the include option, to assert that only the provided collection are being kept in the file `packages/datasource-customizer/test/datasource-customizer.test.ts`.

File: packages/datasource-customizer/src/datasource-customizer.ts (hunk 3 above)
Comment (romain-gilliotte) [reply packages/datasource-customizer/src/datasource-customizer.ts:43]: I renamed it to `keepCollectionsMatching`
Publicate does not exists in english... that's badly translated french

File: packages/datasource-customizer/src/decorators/publication-collection/datasource.ts (hunk 5 above)
Comment (romain-gilliotte) [reply packages/datasource-customizer/src/decorators/publication-collection/datasource.ts:20]: Yep. Not super useful though.

`addDataSource(..., { include: ['steve', 'romain'], exclude: ['romain'] })`
=> Keep only steve

In the initial spec we did months ago we wanted to support regexp matching, which would make a lot more sense but I did not implement it because I got lazy.

`addDataSource(..., { include: [/^payfit_/], exclude: ['payfit_credit_cards'] })`
=> Keep all collection prefixed with `payfit_` but not 'payfit_credit_cards'

File: packages/datasource-customizer/src/decorators/publication-collection/datasource.ts (hunk 2 above)
Comment (romain-gilliotte) [reply packages/datasource-customizer/src/decorators/publication-collection/datasource.ts:29]: It's more complicated than that, we can do a meet.

There is a different schema for each layer in the decorator stack.
This tells that all layers on top of us won't have that field in the schema

File: packages/datasource-customizer/test/decorators/publication-collection/datasource.test.ts (hunk 6 above)
Comment (romain-gilliotte) [reply packages/datasource-customizer/test/decorators/publication-collection/datasource.test.ts:95]: good catch

File: docs/datasources/connection/relationships.md (hunk 4 above)
Comment (romain-gilliotte) [reply docs/datasources/connection/relationships.md:1]: It was intended but you're right it is not really related.

My issue was that this PR was introducing a new option to `addDataSource`, and that
- all documented options were under the `Multiple Datasources` section.
- this new option is not strictly related to customer in that situation

=> I refactored this part of the doc

File: packages/datasource-customizer/src/decorators/publication-collection/datasource.ts (hunk 2 above)
Comment (romain-gilliotte) [reply packages/datasource-customizer/src/decorators/publication-collection/datasource.ts:29]: if you look at the `delete` call on line 35, it's actually the same:
we're removing the field only from our layer (we're deleting in `this`).

So underlying layers don't know

File: docs/datasources/connection/partial-imports.md [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+const agent = createAgent(options);
+const aDataSource = createSqlDataSource('postgres://user:pass@localhost:5432/mySchema');
+const anoterDataSource = createSqlDataSource('postgres://user:pass@localhost:5432/mySchema');
+
+// Rename sqlDataSource collections by providing replacements
+agent
+  .addDataSource(aDataSource, { include: ['books', 'reviews'] })
+  .addDataSource(anoterDataSource, { exclude: ['authors'] });
+```
```
[the remark this answers is not in the document]
Comment (romain-gilliotte) [reply docs/datasources/connection/partial-imports.md]: 😱

File: docs/datasources/connection/relationships.md (hunk 4 above)
Comment (SteveBunlon) [reply docs/datasources/connection/relationships.md:1]: Roger that.

File: packages/datasource-customizer/src/decorators/publication-collection/datasource.ts (hunk 5 above)
Comment (SteveBunlon) [reply packages/datasource-customizer/src/decorators/publication-collection/datasource.ts:20]: Let's keep it that way, the regexp way serves edge cases, that can still be achieved by only providing static include exclude. KISS.

**Review by SteveBunlon — APPROVED** [review]
Tested on the test agent, works like a charm ❇️
